File: 90-subsets-ii/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        def backtracking(i,subset):
            if(i==len(nums)): # curr_state==sol_state
                res.append(subset[:]) # adding a copy of subset
                return
            subset.append(nums[i]) # make the move # pick
            backtracking(i+1,subset) # backtrack
            while(i+1<len(nums) and nums[i]==nums[i+1]):
                i=i+1
            subset.remove(nums[i]) # undo the move  # noty pick
            backtracking(i+1,subset) # backtrack 
       #----------------------------------------
        res=[]
        nums.sort()
        backtracking(0,[]) # i=0 subset=[]
        return res

File: 90-subsets-ii/subsets-ii.java
class Solution{
public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res= new ArrayList<>();
        ArrayList<Integer> subset= new ArrayList<>();
        backtracking(nums,0,subset,res);
        return res;
    }
    private void backtracking(int[] nums, int i, ArrayList<Integer> subset, List<List<Integer>> res){

        if(i==nums.length){
            res.add(new ArrayList<>(subset));// adding a copy of subset
            return;
        }
        subset.add(nums[i]); // Pick making a choice
        backtracking(nums,i+1,subset,res); // exploring a path
        subset.remove(subset.size()-1);// not picking # undo the choice
        while(i<nums.length-1 && nums[i]==nums[i+1]){
            i++;
        }
        backtracking(nums,i+1,subset,res); // exploring a path
    }
}